domain/src/main/java/org/oppia/domain/oppialogger/exceptions/ExceptionsController.kt

```diff
@@ -0,0 +1,150 @@
+   */
+  private fun uploadOrCacheExceptionLog(
+    exception: Exception,
+    timestamp: Long,
+    exceptionType: ExceptionLog.ExceptionType
+  ) {
+    when (networkConnectionUtil.getCurrentConnectionStatus()) {
+      NetworkConnectionUtil.ConnectionStatus.NONE ->
+        cacheExceptionLog(convertExceptionToExceptionLog(exception, timestamp, exceptionType))
+      else -> exceptionLogger.logException(exception)
+    }
+  }
+
+  /** Returns an [ExceptionLog] from an [throwable]. */
+  private fun convertExceptionToExceptionLog(
+    throwable: Throwable,
+    timestamp: Long,
+    exceptionType: ExceptionLog.ExceptionType
+  ): ExceptionLog {
+    val exceptionLogBuilder = ExceptionLog.newBuilder()
+    throwable.message?.let {
+      exceptionLogBuilder.message = it
+    }
+    exceptionLogBuilder.timestampInMillis = timestamp
+    throwable.cause?.let {
+      exceptionLogBuilder.cause = convertExceptionToExceptionLog(it, timestamp, exceptionType)
+    }
+    throwable.stackTrace?.let {
+      exceptionLogBuilder.addAllStacktraceElement(it.map(this::convertStackTraceElementToLog))
+    }
+    exceptionLogBuilder.exceptionType = exceptionType
+    return exceptionLogBuilder.build()
+  }
+
+/** Builds the [ExceptionLog.StackTraceElement] from a [stackTraceElement]. */
+  private fun convertStackTraceElementToLog(
+    stackTraceElement: StackTraceElement
+  ): ExceptionLog.StackTraceElement {
+    return ExceptionLog.StackTraceElement.newBuilder()
+      .setFileName(stackTraceElement.fileName)
+      .setMethodName(stackTraceElement.methodName)
+      .setLineNumber(stackTraceElement.lineNumber)
+      .setDeclaringClass(stackTraceElement.className)
+      .build()
+  }
+
+  /**
+   * Adds an exception to the storage.
+   *
+   * At first, it checks if the size of the store isn't exceeding [exceptionLogStorageCacheSize].
+   * If the limit is exceeded then the least recent exception is removed from the [exceptionLogStore].
+   * After this, the [exceptionLog] is added to the store.
+   * */
+  private fun cacheExceptionLog(exceptionLog: ExceptionLog) {
+    exceptionLogStore.storeDataAsync(true) { oppiaExceptionLogs ->
+      val storeSize = oppiaExceptionLogs.exceptionLogList.size
+      if (storeSize + 1 > exceptionLogStorageCacheSize) {
+        val exceptionLogRemovalIndex = getLeastRecentExceptionIndex(oppiaExceptionLogs)
+        if (exceptionLogRemovalIndex != null) {
+          return@storeDataAsync oppiaExceptionLogs.toBuilder()
+            .removeExceptionLog(exceptionLogRemovalIndex)
+            .addExceptionLog(exceptionLog)
+            .build()
+        } else {
+          // TODO (#1433): Refactoring for logging exceptions to both console and exception loggers.
+          val exception =
+            NullPointerException(
+              "Least Recent Exception index absent -- ExceptionLogCacheStoreSize is 0"
+            )
+          consoleLogger.e("Exceptions Controller", exception.toString())
```

Comment:
Can we put this into a constant and use it in the file?

Reply:
Done.

Reply:
Done.